design/proposal-generic-log.md

```diff
@@ -0,0 +1,150 @@
+
+Example code broadly demonstrating the design:
+
+```go
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type collector interface {
+	Collect(chan []string) string
+}
+
+type someCollector struct {
+	pods []string
+}
+
+func (c *someCollector) Collect(ch chan []string) string {
+
+	fmt.Println("running collect on someCollector")
+	time.Sleep(1 * time.Second)
+	ch <- c.pods
+	return "someCollector finished"
+}
+
+type cephCollector struct {
+	cephpods []string
+}
+
+func (c *cephCollector) Collect(ch chan []string) string {
+	fmt.Println("running collect on cephCollector")
+	time.Sleep(2 * time.Second)
+	ch <- c.cephpods
+	return "cephCollector finished"
+}
+
+type podLogCollector struct {
+}
+
+// function that receives the list of pods to collect logs from
+// alter the logs collector to just send a list here, this is a new function
+// to go collect the actual logs.
+func (p *podLogCollector) GetPodLogs(ch chan []string) string {
+	// consolidate all the pods and dedupe
+	for pod := range ch {
+		time.Sleep(1 * time.Second)
+		fmt.Println("podLogCollector collected logs from", pod)
+		// add the list to an index
+		// if the logs haven't already been collected, go fetch them
+	}
+	return "podLogCollector finished"
+}
+
+func main() {
+	podChan := make(chan []string, 2)
+	collectors := []collector{
+		&someCollector{pods: []string{"pod1", "pod2"}},
+		&cephCollector{cephpods: []string{"rook1", "osd1", "mon1"}},
+		&someCollector{pods: []string{"pod3", "pod4"}},
+		&someCollector{pods: []string{"pod5", "pod6"}},
+	}
+
+	var logsWg sync.WaitGroup
+	logsWg.Add(1)
+	// grab a list of pods from which to collect logs via a channel read
+	logs := &podLogCollector{}
+	go func(l *podLogCollector) {
+		fmt.Println("Running GetPodLogs.... ", l.GetPodLogs(podChan))
+		defer logsWg.Done()
+	}(logs)
+
+	// run the collectors, they need to send the pod list which wants logs to the channel
+	var wg sync.WaitGroup
+	for _, coll := range collectors {
+		wg.Add(1)
+		go func(coll collector) {
+			defer wg.Done()
+			fmt.Println(coll.Collect(podChan))
+		}(coll)
+	}
+
+	wg.Wait()
+	close(podChan)
+	logsWg.Wait()
+}
+```
+
+## Limitations
+ 
+## Assumptions
+
+* The Kubernetes API can handle the load of collectors running concurrently, including when the pod logs collector runs
+ 
+## Testing
+
+The new function will need unit tests.
+
+The existing Collect functions will need unit tests altered (or added).
+
+## Documentation
+
+No changes required.
+
+## Alternatives Considered
+
+* Having each collector append the list of pods to a global, then on completion of all collectors run the collector for pod logs (starting with deduplicating the list of pods).  This would be slower since we need to wait for all collectors to complete prior to pod log collection starting.
+* Making a generic logs collector function that can be imported by any collector that wishes to collect logs.  This would not allow the same pod's logs to be collected several times.
```

Comment:
Not sure I understand `This would not allow the same pod's logs to be collected several times.`.

Reply:
need to express it better in the text - the idea is that the logs collection function keeps a list of pods it has already collected, and ignores requests for pods it's already grabbed.

Reply:
I've updated the statement, but I'm not sure when looking at these alternatives - maybe a common log collector function that others import is a simpler option rather than messing with channels etc, and a better option therefore.